core/src/main/java/feast/core/service/JobService.java

```diff
@@ -0,0 +1,282 @@
+        if (filter.getStoreName() != "") {
+          // find jobs by name
+          List<Job> jobs = this.jobRepository.findByStoreName(filter.getStoreName());
+          Set<String> jobIds = jobs.stream().map(Job::getId).collect(Collectors.toSet());
+          matchingJobIds = this.mergeResults(matchingJobIds, jobIds);
+        }
+        if (filter.hasFeatureSetReference()) {
+          // find a matching featuresets for reference
+          FeatureSetReference fsReference = filter.getFeatureSetReference();
+          ListFeatureSetsResponse response =
+              this.specService.listFeatureSets(this.toListFeatureSetFilter(fsReference));
+          List<FeatureSet> featureSets =
+              response.getFeatureSetsList().stream()
+                  .map(FeatureSet::fromProto)
+                  .collect(Collectors.toList());
+
+          // find jobs for the matching featuresets
+          Collection<Job> matchingJobs = this.jobRepository.findByFeatureSetsIn(featureSets);
+          List<String> jobIds = matchingJobs.stream().map(Job::getId).collect(Collectors.toList());
+          matchingJobIds = this.mergeResults(matchingJobIds, jobIds);
+        }
+      }
+    } else {
+      // no or empty filter: match all jobs
+      matchingJobIds =
+          this.jobRepository.findAll().stream().map(Job::getId).collect(Collectors.toSet());
+    }
+
+    // convert matching job models to ingestion job protos
+    List<IngestionJobProto.IngestionJob> ingestJobs = new ArrayList<>();
+    for (String jobId : matchingJobIds) {
+      Job job = this.jobRepository.findById(jobId).get();
+      ingestJobs.add(job.toProto());
+    }
+
+    // pack jobs into response
+    return ListIngestionJobsResponse.newBuilder().addAllJobs(ingestJobs).build();
+  }
+
+  /**
+   * Restart (Aborts) the ingestion job matching the given restart request. See CoreService protobuf
+   * documentation for more detailed documentation.
+   *
+   * @param request restart ingestion job request specifying which job to stop
+   * @throws NoSuchElementException when restart job request requests to restart a nonexistent job.
+   * @throws UnsupportedOperationException when job to be restarted is in an unsupported status
+   * @throws InvalidProtocolBufferException on error when constructing response protobuf
+   */
+  @Transactional
+  public RestartIngestionJobResponse restartJob(RestartIngestionJobRequest request)
+      throws InvalidProtocolBufferException {
+    // check job exists
+    Optional<Job> getJob = this.jobRepository.findById(request.getId());
+    if (getJob.isEmpty()) {
+      throw new NoSuchElementException(
+          "Attempted to stop nonexistent job with id: " + getJob.get().getId());
+    }
+
+    // check job status is valid for restarting
+    Job job = getJob.get();
+    JobStatus status = job.getStatus();
+    if (JobStatus.getTransitionalStates().contains(status) || status.equals(JobStatus.UNKNOWN)) {
```

Comment:
It might make sense to disallow restarting of jobs in terminal state as well, in order to prevent race conditions with the job coordinator. Another (smaller) issue with allowing job restarts for terminal jobs is that Dataflow only retains information about stopped jobs for a short period of time, after which the job cannot be retrieved any more.

Reply:
Thanks for pointing the race condition out, I have disabled restarting for terminal job statuses. As for your second point, this would not be a problem as the Dataflow Job Manager is configured to recreate the job once stopped.